From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lawrence Chen <lawrence@manaflow.com>
Date: Mon, 13 Jul 2026 11:26:02 +0000
Subject: fix: preserve ProcessSingleton data in Chrome runtime


diff --git a/chrome/browser/chrome_browser_main.cc b/chrome/browser/chrome_browser_main.cc
index 570f788355f25cdd22ef86f5974c89efce76ce43..518547d0fc6b7bee5ae55f398c530ef75ae0a6c7 100644
--- a/chrome/browser/chrome_browser_main.cc
+++ b/chrome/browser/chrome_browser_main.cc
@@ -524,7 +524,8 @@ void ProcessSingletonNotificationCallbackImpl(
 // (i.e., because the current process is shutting down).
 bool ProcessSingletonNotificationCallback(
     base::CommandLine command_line,
-    const base::FilePath& current_directory) {
+    const base::FilePath& current_directory,
+    std::vector<uint8_t> /*additional_data*/) {
   // Drop the request if the browser process is already shutting down.
   // Note that we're going to post an async task below. Even if the browser
   // process isn't shutting down right now, it could be by the time the task
@@ -2302,6 +2303,7 @@ std::unique_ptr<base::RunLoop> ChromeBrowserMainParts::TakeRunLoopForTest() {
 bool ChromeBrowserMainParts::ProcessSingletonNotificationForTesting(
     base::CommandLine command_line) {
   return ProcessSingletonNotificationCallback(command_line,
-                                              /*current_directory=*/{});
+                                              /*current_directory=*/{},
+                                              /*additional_data=*/{});
 }
 #endif  // BUILDFLAG(ENABLE_PROCESS_SINGLETON)
diff --git a/chrome/browser/chrome_process_singleton.cc b/chrome/browser/chrome_process_singleton.cc
index 4e1b0de17710092074a7eb68bc6260eb5edfb66d..87a52658d7ef9dc2c469a2402e6084f509ab3f6d 100644
--- a/chrome/browser/chrome_process_singleton.cc
+++ b/chrome/browser/chrome_process_singleton.cc
@@ -28,8 +28,15 @@ ChromeProcessSingleton::ChromeProcessSingleton(
     : startup_lock_(
           base::BindRepeating(&ChromeProcessSingleton::NotificationCallback,
                               base::Unretained(this))),
-      process_singleton_(user_data_dir,
-                         startup_lock_.AsNotificationCallback()) {}
+      process_singleton_(
+#if BUILDFLAG(IS_WIN)
+          "Chrome",
+#endif
+          user_data_dir, {},
+#if BUILDFLAG(IS_WIN)
+          /*is_app_sandboxed=*/false,
+#endif
+          startup_lock_.AsNotificationCallback()) {}
 
 ChromeProcessSingleton::~ChromeProcessSingleton() = default;
 
@@ -137,7 +144,9 @@ bool ChromeProcessSingleton::IsSingletonInstance() {
 
 bool ChromeProcessSingleton::NotificationCallback(
     base::CommandLine command_line,
-    const base::FilePath& current_directory) {
+    const base::FilePath& current_directory,
+    std::vector<uint8_t> additional_data) {
   DCHECK(notification_callback_);
-  return notification_callback_.Run(std::move(command_line), current_directory);
+  return notification_callback_.Run(std::move(command_line), current_directory,
+                                    std::move(additional_data));
 }
diff --git a/chrome/browser/chrome_process_singleton.h b/chrome/browser/chrome_process_singleton.h
index dc79446b281fc2d08b6cc13ff266bc77d63c3afe..a7f610a817d9d06c2481ef29fa49600d1486ba33 100644
--- a/chrome/browser/chrome_process_singleton.h
+++ b/chrome/browser/chrome_process_singleton.h
@@ -5,6 +5,9 @@
 #ifndef CHROME_BROWSER_CHROME_PROCESS_SINGLETON_H_
 #define CHROME_BROWSER_CHROME_PROCESS_SINGLETON_H_
 
+#include <cstdint>
+#include <vector>
+
 #include "base/files/file_path.h"
 #include "base/functional/callback.h"
 #include "chrome/browser/process_singleton.h"
@@ -68,7 +71,8 @@ class ChromeProcessSingleton {
 
  private:
   bool NotificationCallback(base::CommandLine command_line,
-                            const base::FilePath& current_directory);
+                            const base::FilePath& current_directory,
+                            std::vector<uint8_t> additional_data);
 
   // Whether or not this instance is the running single instance.
   bool is_singleton_instance_ = false;
diff --git a/chrome/browser/chrome_process_singleton_win_unittest.cc b/chrome/browser/chrome_process_singleton_win_unittest.cc
index 016fa9db7d95028c094a032df3daa5afc50c59c6..ad2a91e425dbc84abdc6d6b91eb44fcb3b65645d 100644
--- a/chrome/browser/chrome_process_singleton_win_unittest.cc
+++ b/chrome/browser/chrome_process_singleton_win_unittest.cc
@@ -8,6 +8,9 @@
 
 #include <processthreadsapi.h>
 
+#include <cstdint>
+#include <vector>
+
 #include "base/command_line.h"
 #include "base/compiler_specific.h"
 #include "base/files/file_path.h"
@@ -24,13 +27,15 @@ namespace {
 
 bool ServerCallback(int* callback_count,
                     base::CommandLine command_line,
-                    const base::FilePath& current_directory) {
+                    const base::FilePath& current_directory,
+                    std::vector<uint8_t> additional_data) {
   ++(*callback_count);
   return true;
 }
 
 bool ClientCallback(base::CommandLine command_line,
-                    const base::FilePath& current_directory) {
+                    const base::FilePath& current_directory,
+                    std::vector<uint8_t> additional_data) {
   ADD_FAILURE();
   return false;
 }
diff --git a/chrome/browser/process_singleton_posix_unittest.cc b/chrome/browser/process_singleton_posix_unittest.cc
index 977bc6b7ee4e45dbd39de2576f2b65c00b9c8bce..5c683f74effbcf7bfc2443a168008938d47f4687 100644
--- a/chrome/browser/process_singleton_posix_unittest.cc
+++ b/chrome/browser/process_singleton_posix_unittest.cc
@@ -51,7 +51,7 @@ class ProcessSingletonPosixTest : public testing::Test {
   class TestableProcessSingleton : public ProcessSingleton {
    public:
     explicit TestableProcessSingleton(const base::FilePath& user_data_dir)
-        : ProcessSingleton(user_data_dir,
+        : ProcessSingleton(user_data_dir, {},
                            base::BindRepeating(
                                &TestableProcessSingleton::NotificationCallback,
                                base::Unretained(this))) {}
@@ -66,7 +66,8 @@ class ProcessSingletonPosixTest : public testing::Test {
 
    private:
     bool NotificationCallback(base::CommandLine command_line,
-                              const base::FilePath& current_directory) {
+                              const base::FilePath& current_directory,
+                              std::vector<uint8_t> additional_data) {
       callback_command_lines_.push_back(std::move(command_line));
       return true;
     }
diff --git a/chrome/browser/process_singleton_startup_lock.cc b/chrome/browser/process_singleton_startup_lock.cc
index 6b64b9f72aa93e9f9bf22e1601a0f79d7aff8144..50bec1ddce755d43088cadddafea1e8edeb25240 100644
--- a/chrome/browser/process_singleton_startup_lock.cc
+++ b/chrome/browser/process_singleton_startup_lock.cc
@@ -35,7 +35,8 @@ void ProcessSingletonStartupLock::Unlock() {
        it != saved_startup_messages_.end(); ++it) {
     if (replayed_messages.find(*it) != replayed_messages.end())
       continue;
-    original_callback_.Run(base::CommandLine(it->first), it->second);
+    original_callback_.Run(base::CommandLine(std::get<0>(*it)),
+                           std::get<1>(*it), std::get<2>(*it));
     replayed_messages.insert(*it);
   }
   saved_startup_messages_.clear();
@@ -43,14 +44,16 @@ void ProcessSingletonStartupLock::Unlock() {
 
 bool ProcessSingletonStartupLock::NotificationCallbackImpl(
     base::CommandLine command_line,
-    const base::FilePath& current_directory) {
+    const base::FilePath& current_directory,
+    std::vector<uint8_t> additional_data) {
   if (locked_) {
     // If locked, it means we are not ready to process this message because
     // we are probably in a first run critical phase.
-    saved_startup_messages_.push_back(
-        std::make_pair(command_line.argv(), current_directory));
+    saved_startup_messages_.emplace_back(
+        command_line.argv(), current_directory, std::move(additional_data));
     return true;
   } else {
-    return original_callback_.Run(std::move(command_line), current_directory);
+    return original_callback_.Run(std::move(command_line), current_directory,
+                                  std::move(additional_data));
   }
 }
diff --git a/chrome/browser/process_singleton_startup_lock.h b/chrome/browser/process_singleton_startup_lock.h
index 7a28b2e4189a20f41bb99433ee6614b98630bee6..20569b7740204d6343030b3c688aac9798626ee2 100644
--- a/chrome/browser/process_singleton_startup_lock.h
+++ b/chrome/browser/process_singleton_startup_lock.h
@@ -5,8 +5,9 @@
 #ifndef CHROME_BROWSER_PROCESS_SINGLETON_STARTUP_LOCK_H_
 #define CHROME_BROWSER_PROCESS_SINGLETON_STARTUP_LOCK_H_
 
+#include <cstdint>
 #include <set>
-#include <utility>
+#include <tuple>
 #include <vector>
 
 #include "base/command_line.h"
@@ -45,11 +46,14 @@ class ProcessSingletonStartupLock {
   bool locked() { return locked_; }
 
  private:
-  typedef std::pair<base::CommandLine::StringVector, base::FilePath>
-      DelayedStartupMessage;
+  using DelayedStartupMessage =
+      std::tuple<base::CommandLine::StringVector,
+                 base::FilePath,
+                 std::vector<uint8_t>>;
 
   bool NotificationCallbackImpl(base::CommandLine command_line,
-                                const base::FilePath& current_directory);
+                                const base::FilePath& current_directory,
+                                std::vector<uint8_t> additional_data);
 
   bool locked_;
   std::vector<DelayedStartupMessage> saved_startup_messages_;
diff --git a/chrome/browser/process_singleton_win_unittest.cc b/chrome/browser/process_singleton_win_unittest.cc
index 23c44186cf813d4bbd8f54efa8c6d0761e0941f7..92e1285e0c39a18fde3410adca934cd3375e912f 100644
--- a/chrome/browser/process_singleton_win_unittest.cc
+++ b/chrome/browser/process_singleton_win_unittest.cc
@@ -6,8 +6,10 @@
 
 #include <windows.h>
 
+#include <cstdint>
 #include <memory>
 #include <string>
+#include <vector>
 
 #include "base/check.h"
 #include "base/command_line.h"
@@ -47,7 +49,8 @@ const int kErrorResultCode = 0x345;
 const char kLockfile[] = "lockfile";
 
 bool NotificationCallback(base::CommandLine command_line,
-                          const base::FilePath& current_directory) {
+                          const base::FilePath& current_directory,
+                          std::vector<uint8_t> additional_data) {
   // This is never called in this test, but would signal that the singleton
   // notification was successfully handled.
   NOTREACHED();
@@ -137,7 +140,8 @@ MULTIPROCESS_TEST_MAIN(ProcessSingletonTestProcessMain) {
 
   // Instantiate the process singleton.
   ProcessSingleton process_singleton(
-      user_data_dir, base::BindRepeating(&NotificationCallback));
+      "Chrome", user_data_dir, {}, /*is_app_sandboxed=*/false,
+      base::BindRepeating(&NotificationCallback));
 
   if (!process_singleton.Create())
     return kErrorResultCode;
@@ -240,7 +244,8 @@ class ProcessSingletonTest : public base::MultiProcessTest {
     // The ready event has been signalled - the process singleton is held by
     // the hung sub process.
     test_singleton_ = std::make_unique<ProcessSingleton>(
-        user_data_dir(), base::BindRepeating(&NotificationCallback));
+        "Chrome", user_data_dir(), {}, /*is_app_sandboxed=*/false,
+        base::BindRepeating(&NotificationCallback));
 
     test_singleton_->OverrideShouldKillRemoteProcessCallbackForTesting(
         base::BindRepeating(&ProcessSingletonTest::MockShouldKillRemoteProcess,
@@ -382,8 +387,9 @@ TEST_F(ProcessSingletonTest, DeterministicDestructionOrder) {
   ASSERT_TRUE(profile_dir.CreateUniqueTempDir());
 
   // Initialize ProcessSingleton directly with base::NullCallback()
-  auto ps = std::make_unique<ProcessSingleton>(profile_dir.GetPath(),
-                                               base::NullCallback());
+  auto ps = std::make_unique<ProcessSingleton>(
+      "Chrome", profile_dir.GetPath(), {}, /*is_app_sandboxed=*/false,
+      base::NullCallback());
 
   // Acquire the lock (we become the master)
   ASSERT_TRUE(ps->Create());
